source/common/common/logger.cc

```diff
@@ -19,18 +19,32 @@ namespace Logger {
 StandardLogger::StandardLogger(const std::string& name)
     : Logger(std::make_shared<spdlog::logger>(name, Registry::getSink())) {}
 
-SinkDelegate::SinkDelegate(DelegatingLogSinkSharedPtr log_sink)
-    : previous_delegate_(log_sink->delegate()), log_sink_(log_sink) {
-  log_sink->setDelegate(this);
-}
+SinkDelegate::SinkDelegate(DelegatingLogSinkSharedPtr log_sink) : log_sink_(log_sink) {}
 
 SinkDelegate::~SinkDelegate() {
-  assert(log_sink_->delegate() == this); // Ensures stacked allocation of delegates.
+  // The previous delegate should have never been set or should have been reset by now via
+  // restoreDelegate();
+  assert(previous_delegate_ == nullptr);
+}
+
+void SinkDelegate::setDelegate() {
+  // There should be no previous delegate before this call.
+  assert(previous_delegate_ == nullptr);
+  previous_delegate_ = log_sink_->delegate();
+  log_sink_->setDelegate(this);
+}
+
+void SinkDelegate::restoreDelegate() {
+  // Ensures stacked allocation of delegates.
+  assert(log_sink_->delegate() == this);
   log_sink_->setDelegate(previous_delegate_);
+  previous_delegate_ = nullptr;
 }
 
 StderrSinkDelegate::StderrSinkDelegate(DelegatingLogSinkSharedPtr log_sink)
-    : SinkDelegate(log_sink) {}
+    : SinkDelegate(log_sink) {
+  setDelegate();
```

Comment:
I'm missing where restore is called for this one?

Reply:
shouldn't that have failed one of your asserts?  Or can we make it so it would have?

Reply:
I think the stderr sink is never deleted in the common case. Let me add a test.